# pandas

This workshop's goal&mdash;which is facilitated by this Jupyter notebook&mdash;is to give attendees the confidence to use `pandas` in their research projects. Basic familiarity with Python *is* assumed.

`pandas` is designed to make it easier to work with structured data. Most of the analyses you might perform will likely involve using tabular data, e.g., from .csv files or relational databases (e.g., SQL). The `DataFrame` object in `pandas` is "a two-dimensional tabular, column-oriented data structure with both row and column labels."

If you're curious:

>The `pandas` name itself is derived from *panel data*, an econometrics term for multidimensional structured data sets, and *Python data analysis* itself.

To motivate this workshop, we'll work with example data and go through the various steps you might need to prepare data for analysis. You'll (hopefully) realize that doing this type of work is much more difficult using Python's built-in data structures.

The data used in these examples is available in the following [GitHub repository](https://github.com/juanshishido/introduction-to-pandas). If you've [cloned that repo](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-clone), which is the recommended approach, you'll have everything you need to run this notebook. Otherwise, you can download the data file(s) from the above link. (Note: this notebook assumes that the data files are in a directory named `data/` found within your current working directory.)

For this example, we're working with European unemployment data from Eurostat, which is hosted by [Google](https://code.google.com/p/dspl/downloads/list). There are several .csv files that we'll work with in this workshop.

Let's begin by importing `pandas` using the conventional abbreviation. (We'll also import `numpy`, which we'll use for some operations.)

In [1]:
import numpy as np
import pandas as pd

The `read_csv()` function in `pandas` allows us to easily import our data. By default, it assumed the data is comma-delimited. However, you can specify the delimiter used in your data (e.g., tab, semicolon, pipe, etc.). There are several parameters that you can specify. See the documentation [here](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html). `read_csv()` returns a `DataFrame`.

Notice that we call `read_csv()` using the `pd` abbreviation from the import statement above.

In [2]:
unemployment = pd.read_csv('data/country_total.csv')

Great! You've created a `pandas` `DataFrame`. We can look at our data by using the `.head()` method. By default, this shows the header (column names) and the first five rows. Passing an integer, $n$, to `.head()` returns that number of rows. To see the last $n$ rows, use `.tail()`.

In [3]:
unemployment.head()

country seasonality    month  unemployment  unemployment_rate
0      at         nsa  1993.01        171000                4.5
1      at         nsa  1993.02        175000                4.6
2      at         nsa  1993.03        166000                4.4
3      at         nsa  1993.04        157000                4.1
4      at         nsa  1993.05        147000                3.9

To find the number of rows, you can use the `len()` function. Alternatively, you can use the `.shape` attribute.

In [4]:
unemployment.shape

(20796, 5)

There are 20,796 rows and 5 columns.

You may have noticed that the `month` column also includes the year. Let's go ahead and rename it.

In [5]:
unemployment.rename(columns={'month' : 'year_month'}, inplace=True)

The `.rename()` methods allows you to modify index labels and/or column names. As you can see, we passed a `dict` to the `column` parameter, with the original name as the key and the new name as the value. Importantly, we also set the `inplace` parameter to `True`, which modifies the *actual* `DataFrame`, not a copy of it.

It might also make sense to separate the data in `year_month` into two separate columns. To do this, you'll need to know how to select a single column. We can either use bracket (`[]`) or dot notation (referred to as *attribute access*).

In [6]:
unemployment['year_month'].head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

In [7]:
unemployment.year_month.head()

0    1993.01
1    1993.02
2    1993.03
3    1993.04
4    1993.05
Name: year_month, dtype: float64

It is preferrable to use the bracket notation as a column name might inadvertently have the same name as a `DataFrame` (or `Series`) method. In addition, only bracket notation can be used to create a new column. If you try and use attribute access to create a new column, you'll create a new attribute, *not* a new column.

When selecting a single column, we have is a `pandas` `Series` object, which is a single vector of data (e.g., a NumPy array) with "an associated array of data labels, called its *index*." A `DataFrame` also has an index. In our example, the indices are an array of sequential integers, which is the default. You can find them in the left-most position, without a column label.

Indices need not be a sequence of integers. They can, for example, be dates or strings.

Indices, like column names, can be used to select data. Indices can be used to select particular rows. In fact, you can do something like `.head()` with slicing using the `[]` operator.

In [8]:
unemployment[:5]

country seasonality  year_month  unemployment  unemployment_rate
0      at         nsa     1993.01        171000                4.5
1      at         nsa     1993.02        175000                4.6
2      at         nsa     1993.03        166000                4.4
3      at         nsa     1993.04        157000                4.1
4      at         nsa     1993.05        147000                3.9

---

Before we continue, let's look at a few useful ways to index data&mdash;that is, select rows.

`.loc` primarily works with string labels. It accepts a single label, a list (or array) of labels, or a slice of labels (e.g., `'a' : 'f'`).

Let's create a `DataFrame` to see how this works. (This is based on an [example](https://github.com/fonnesbeck/scipy2015_tutorial/blob/master/notebooks/1.%20Data%20Preparation.ipynb) from Chris Fonnesbeck's [Computational Statistics II Tutorial](https://github.com/fonnesbeck/scipy2015_tutorial).

In [9]:
bacteria = pd.DataFrame({'bacteria_counts' : [632, 1638, 569, 115],
                         'other_feature' : [438, 833, 234, 298]},
                         index=['Firmicutes', 'Proteobacteria', 'Actinobacteria', 'Bacteroidetes'])

In [10]:
bacteria

bacteria_counts  other_feature
Firmicutes                  632            438
Proteobacteria             1638            833
Actinobacteria              569            234
Bacteroidetes               115            298

Now, if we're interested in the values associated with "Actinobacteria," we can use `.loc` and the index name.

In [11]:
bacteria.loc['Actinobacteria']

bacteria_counts    569
other_feature      234
Name: Actinobacteria, dtype: int64

This returns the column values for the specified row. Interestingly, we could have used "positional indexing."

In [12]:
bacteria[2:3]

bacteria_counts  other_feature
Actinobacteria              569            234

The difference is that the former returns a `Series`, while the latter returns a `DataFrame`.

Let's return to our unemployment data. Another indexing option, `.iloc`, primarily works with integer positions. It accepts the same kinds of data, but with integers. To select specific rows, we can do the following.

In [13]:
unemployment.iloc[[1, 5, 6, 9]]

country seasonality  year_month  unemployment  unemployment_rate
1      at         nsa     1993.02        175000                4.6
5      at         nsa     1993.06        134000                3.5
6      at         nsa     1993.07        128000                3.4
9      at         nsa     1993.10        141000                3.7

We can select a range of rows, and specify the step value.

In [14]:
unemployment.iloc[25:50:5]

country seasonality  year_month  unemployment  unemployment_rate
25      at         nsa     1995.02        174000                4.5
30      at         nsa     1995.07        123000                3.3
35      at         nsa     1995.12        175000                4.7
40      at         nsa     1996.05        159000                4.3
45      at         nsa     1996.10        146000                3.9

(Note: As is typical in Python, the end position is not included. Therefore, we don't see the row associated with the index 50.)

Indexing is important. You'll use it a lot. Below, we'll show how to index based on data values.

---

So, we still want to split `year_month` into two separate columns. Above, we saw that this column is type `float64`. We'll first extract the year using NumPy's `floor()` function. We'll then subtract this value from `year_month` and multiply the result by 100.

In [15]:
unemployment['year'] = np.floor(unemployment['year_month'])

In [16]:
unemployment['month'] = (unemployment['year_month'] - unemployment['year']) * 100

To create the `month` column, we subtracted two vectors. This resulted in the decimal value in `year_month`. To transform the values to integers, we multiplied by 100.

Now, let's say we wanted to reorder the columns in the `DataFrame`. For this, we use bracket notation again, passing in a list of column names in the order we'd like to see them.

In [17]:
unemployment = unemployment[['country', 'seasonality',
                             'year_month', 'year', 'month',
                             'unemployment', 'unemployment_rate']]

In [18]:
unemployment.head(10)

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   
5      at         nsa     1993.06  1993      6        134000   
6      at         nsa     1993.07  1993      7        128000   
7      at         nsa     1993.08  1993      8        130000   
8      at         nsa     1993.09  1993      9        132000   
9      at         nsa     1993.10  1993     10        141000   

   unemployment_rate  
0                4.5  
1                4.6  
2                4.4  
3                4.1  
4                3.9  
5                3.5  
6                3.4  
7                3.4  
8                3.5  
9                3.7

So far, our `DataFrame` is organized in a reasonable way. But, we know we can do better. We're eventually going to be interested in the unemployment_rate in for each country. The trouble is, we don't exactly know what the values in `country` refer to. We can fix that by getting country names. These can be found in `countries.csv`. However, instead of loading the file in `data/`, why not try something else?

In [19]:
countries_url = 'https://raw.githubusercontent.com/dlab-berkeley/introduction-to-pandas/master/data/countries.csv'
countries = pd.read_csv(countries_url)

`read_csv()` can take a URL for the file path.

In [20]:
countries.tail(3)

country google_country_code country_group         name_en      name_fr  \
27      se                  SE            eu          Sweden        Suède   
28      tr                  TR        non-eu          Turkey      Turquie   
29      uk                  GB            eu  United Kingdom  Royaume-Uni   

                   name_de   latitude  longitude  
27                Schweden  62.198468  14.896307  
28                  Türkei  38.952942  35.439795  
29  Vereinigtes Königreich  54.315447  -2.232612

This file has lots of useful information. It even has the country names is three different languages.

Because the data we need is stored in two separate files, we'll want to merge the data somehow. Let's determine which column we can use to join this data. `country` looks like a good option. However, we don't need all of the columns in this `DataFrame`. To select certain columns, we use the name bracket notation we used to reorder the columns.

In [21]:
country_names = countries[['country', 'country_group', 'name_en']]

In [22]:
country_names.head(2)

country country_group  name_en
0      at            eu  Austria
1      be            eu  Belgium

`pandas` includes an easy-to-use merge function.

In [23]:
unemployment = pd.merge(unemployment, country_names, on='country')

In [24]:
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

That's better. We now know that the abbreviation "at" corresponds to Austria. We might be curious to check what countries we have data for. The `Series` object includes a `unique()` method. We'll use this to check the countries. We can select the name either using bracket or dot notation.

In [25]:
unemployment.name_en.unique()

array(['Austria', 'Belgium', 'Bulgaria', 'Cyprus', 'Czech Republic',
       'Germany (including  former GDR from 1991)', 'Denmark', 'Estonia',
       'Spain', 'Finland', 'France', 'Greece', 'Croatia', 'Hungary',
       'Ireland', 'Italy', 'Lithuania', 'Luxembourg', 'Latvia', 'Malta',
       'Netherlands', 'Norway', 'Poland', 'Portugal', 'Romania', 'Sweden',
       'Slovenia', 'Slovakia', 'Turkey', 'United Kingdom'], dtype=object)

To get a count of the number of unique countries, wrap the above code with `len()`.

In [26]:
len(unemployment.name_en.unique())

30

It might be more interesting to know how many observations we actually have. For this, we'll introduce `pandas`'s `groupby()` method.

In [27]:
unemployment.groupby('name_en')['name_en'].count()

name_en
Austria                                       648
Belgium                                      1008
Bulgaria                                      576
Croatia                                       324
Cyprus                                        396
Czech Republic                                468
Denmark                                      1008
Estonia                                       387
Finland                                       828
France                                       1008
Germany (including  former GDR from 1991)     504
Greece                                        450
Hungary                                       576
Ireland                                      1008
Italy                                         924
Latvia                                        459
Lithuania                                     459
Luxembourg                                   1008
Malta                                         576
Netherlands                               

Let's explain what just happened. We start with our `DataFrame`. We tell `pandas` that we want to group the data by country name&mdash;that's what goes in the parentheses. Next, we need to tell it what column we'd like to perform the `.count()` operation on. In this case, it's country name again.

This will be useful for our analysis. The maximum number of observations for a given country for this time period is 1,008 observations. We'll note that certain countries, such as Turkey, have far less data.

How about finding the date range for this data set? Let's look at the minimum and maximum years.

In [28]:
unemployment['year'].min(), unemployment['year'].max()

(1983.0, 2010.0)

Next, we should pause for a moment and think about what data we really care about. For our purposes, the variable of interest is `unemployment_rate`. The number of observations by country only reflect the number of instances for each country name. It is possible, maybe even expected, to have some missing data. Let's find out how many unemployment rate values are missing.

In [29]:
unemployment['unemployment_rate'].isnull().sum()

945

The `.isnull()` attribute returns a corresponding boolean value for each entry in the unemployment rate `Series`. In Python `True` is equivalent to 1 and `False` is equivalent to 0. Thus, when we add the result (with `sum()`), we get a count for the total number of missing values.

What if we'd like to know how many missing values exist at the country level? We can take the main part of what we had above and create a new column in the `DataFrame`.

In [30]:
unemployment['unemployment_rate_null'] = unemployment['unemployment_rate'].isnull()

Next, we can use the GroupBy functionality. Because the column we just created is an array of boolean values, we want to add the values.

In [31]:
unemployment.groupby('name_en')['unemployment_rate_null'].sum()

name_en
Austria                                        0
Belgium                                        0
Bulgaria                                     180
Croatia                                      216
Cyprus                                         0
Czech Republic                                 0
Denmark                                        0
Estonia                                        0
Finland                                        0
France                                         0
Germany (including  former GDR from 1991)      0
Greece                                         0
Hungary                                       36
Ireland                                        0
Italy                                          0
Latvia                                         0
Lithuania                                      0
Luxembourg                                     0
Malta                                        180
Netherlands                                    0
Norway      

As we saw above, the number of records for each country differs. We might, then, want to have the missing values by country shown as percentages. Let's create a new `DataFrame` for this.

We'll take the code from above and set the `as_index` parameter to `False`.

In [32]:
unemployment_rate = unemployment.groupby('name_en', as_index=False)['unemployment_rate_null'].sum()

`unemployment_rate` is a `DataFrame` with the information from cell 30, above. It's important to note that using `as_index=False` in `.groupby()` only works if the grouping column(s) are not the same as the columns on which we're performing the operation.

Also, to group by several columns, simply pass in a list of column names to `.groupby()`. For example,

```
unemployment.groupby(['name_en', 'seasonality'])['unemployment_rate'].mean()
```.

Now, let's add the number of observations by country to the `DataFrame`.

In [33]:
unemployment_rate['n_obs'] = unemployment.groupby('name_en')['name_en'].count().values

Here, we need to use the `values` attribute to get an array of the counts. Excluding `values` will result in a column full of `NaN`s. This is because the index in `unemployment.groupby('name_en')['name_en'].count()` is a list of the country names. When creating a new column, `pandas` tries to match on index. Recall that the default index values for a `DataFrame` is a sequence of integers.

Because we know (or have noticed) that the `.groupby()` function returns the values in alphabetical order, we can simply set the new column to the list of values, as we have done. You can, however, be more explicit and create another `DataFrame` and merge on country name.

Finally, let's create the column for the percentage of missing values.

In [34]:
unemployment_rate['null_percentage'] = unemployment_rate['unemployment_rate_null'] / unemployment_rate['n_obs']

In [35]:
unemployment_rate

name_en  unemployment_rate_null  n_obs  \
0                                     Austria                       0    648   
1                                     Belgium                       0   1008   
2                                    Bulgaria                     180    576   
3                                     Croatia                     216    324   
4                                      Cyprus                       0    396   
5                              Czech Republic                       0    468   
6                                     Denmark                       0   1008   
7                                     Estonia                       0    387   
8                                     Finland                       0    828   
9                                      France                       0   1008   
10  Germany (including  former GDR from 1991)                       0    504   
11                                     Greece                       0    450   
12                                    Hungary                      36    576   
13                                    Ireland                       0   1008   
14                                      Italy                       0    924   
15                                     Latvia                       0    459   
16                                  Lithuania                       0    459   
17                                 Luxembourg                       0   1008   
18                                      Malta                     180    576   
19                                Netherlands                       0   1008   
20                                     Norway                       0    786   
21                                     Poland                      72    576   
22                                   Portugal                       0   1008   
23                                    Romania                       0    423   
24                                   Slovakia                     108    576   
25                                   Slovenia                      36    576   
26                                      Spain                     117   1008   
27                                     Sweden                       0   1008   
28                                     Turkey                       0    210   
29                             United Kingdom                       0   1002   

    null_percentage  
0          0.000000  
1          0.000000  
2          0.312500  
3          0.666667  
4          0.000000  
5          0.000000  
6          0.000000  
7          0.000000  
8          0.000000  
9          0.000000  
10         0.000000  
11         0.000000  
12         0.062500  
13         0.000000  
14         0.000000  
15         0.000000  
16         0.000000  
17         0.000000  
18         0.312500  
19         0.000000  
20         0.000000  
21         0.125000  
22         0.000000  
23         0.000000  
24         0.187500  
25         0.062500  
26         0.116071  
27         0.000000  
28         0.000000  
29         0.000000

This is the first time we've called a `DataFrame` without something like `.head()`. By default, `pandas` prints 60 rows. In this case, because there are only 30 countries, we see the entire `DataFrame`.

As we can see, Croatia has lots of missing data. This `DataFrame` contains useful information&mdash;things to consider&mdash;when analyzing the data.

Let's return to our main `DataFrame`. Now that we have the missing values information in `unemployment_rate`, we can drop the last column we added to `unemployment`.

In [36]:
unemployment.drop('unemployment_rate_null', axis=1, inplace=True)

It's important to specify the `axis` parameter. `axis=1` refers to columns. (`axis=0` refers to rows.) `inplace=True` simply modifies the actual `DataFrame`.

In [37]:
unemployment.head()

country seasonality  year_month  year  month  unemployment  \
0      at         nsa     1993.01  1993      1        171000   
1      at         nsa     1993.02  1993      2        175000   
2      at         nsa     1993.03  1993      3        166000   
3      at         nsa     1993.04  1993      4        157000   
4      at         nsa     1993.05  1993      5        147000   

   unemployment_rate country_group  name_en  
0                4.5            eu  Austria  
1                4.6            eu  Austria  
2                4.4            eu  Austria  
3                4.1            eu  Austria  
4                3.9            eu  Austria

Now that we know about the missing values, we have to deal with them. There are two main options:

* fill the missing values with some other values
* do not use observations with missing values
    * in some cases, we might even want to exclude entire countries for our analysis
    
Because countries with missing unemployment rate data have at least 36 missing values, we'll take the second approach and exclude missing values. To do this, we can use `pandas`' built-in `dropna()` function.